File: TP Evaluatorio_I_Aplicación_de_POO_en_otro_lenguaje._Hoy_tomamos_JAVA_y-PYTHON/C++/Class_abstract/Class_abstract.cpp
#include <iostream>
#include <fstream>
using namespace std;
//Player abstract class
class Player
{
public:
Player()
{
    cout << "The Player constructor was invoked." << endl;
    isOpen = false;
    isPlaying = false;
    setVolume(10);
}
~Player()
{
cout << "The Player destructor was invoked." << endl;
}
//método abstracto o virtual puro:
    virtual void open(string filePath) = 0;
void play()
{
    //@ToDo: por ahora simulamos la reproducción correcta
    if(isOpen) isPlaying = true;
    cout << "The audiofile is playing." << endl;
}
void stop()
{
    //@ToDo: por ahora simulamos la detención correcta
    if(isPlaying) isPlaying = false;
    cout << "The audiofile is stopped." << endl;
}
void setVolume(float value)
{
    volume = value;
    cout << "The volume value is: " << volume << endl;
}
protected:
    bool isOpen;//Protegido para poder ser modificado por la subclase
private:
    bool isPlaying;
    float volume;
};

//VLC concrete class
class VLC : public Player
{
public:
VLC()
{
    cout << "The VLC constructor was invoked." << endl;
}
~VLC()
{
    cout << "The VLC destructor was invoked." << endl;
}
void open(string filePath)
{
    //@ToDo: por ahora simulamos la apertura correcta
    isOpen = true;
    cout << "The audiofile: " << filePath << " is open." << endl;
}
void setPitch(float value)
{
    pitch = value;
    cout << "The pitch value is: " << pitch << endl;
}
private:
    float pitch;
};

//Winamp concrete class
class Winamp : public Player
{
public:
Winamp()
{
    cout << "The Winamp constructor was invoked." << endl;
}
~Winamp()
{
    cout << "The Winamp destructor was invoked." << endl;
}
void open(string filePath)
{
    //@ToDo: por ahora simulamos la apertura correcta
    isOpen = true;
    cout << "The audiofile: " << filePath << " is open." << endl;
}
};

int main()
{
    //NO SE PUEDE CREAR UN OBJETO O INSTANCIA DE UNA CLASE ABSTRACTA:
    //Player player;//NO!
    VLC vlcPlayer;
    cout << endl;
    vlcPlayer.open("./resources/orchestral.ogg");
    vlcPlayer.play();
    vlcPlayer.setVolume(13);
    cout << endl;
    cout << endl;

    Winamp winampPlayer;
    cout << endl;
    winampPlayer.open("./resources/orchestral.ogg");
    winampPlayer.play();
    winampPlayer.setVolume(13);
    cout << endl;
return EXIT_SUCCESS;
}

File: TP Evaluatorio_I_Aplicación_de_POO_en_otro_lenguaje._Hoy_tomamos_JAVA_y-PYTHON/Python/Class_abstract/Class_abstract.py
from abc import ABC, abstractmethod 

#Player abstract class
class Player(ABC):

    def __init__(self):

        print("The Player constructor was invoked.")
        self.isOpen = False;
        self.isPlaying = False;
        self.setVolume(10);

    def __del__(self):

        print("The Player destructor was invoked.")

    #método abstracto o virtual puro:
    def open(self, filePath):
        pass

    def play(self):

        #@ToDo: por ahora simulamos la reproducción correcta
        if(self.isOpen) : self.isPlaying = True;
        print("The audiofile is playing.")

    def stop(self):

        #@ToDo: por ahora simulamos la detención correcta
        if(self.isPlaying) : self.isPlaying = False;
        print("The audiofile is stopped.")

    def setVolume(self, value):

        self.volume = value;
        print("The volume value is: " + str(self.volume))

#VLC concrete class
class VLC(Player):

    def __init__(self):

        print("The VLC constructor was invoked.")

    def __del__(self):

        print("The VLC destructor was invoked.")

    def open(self, filePath):

        #@ToDo: por ahora simulamos la apertura correcta
        self.isOpen = True;
        print("The audiofile: " + filePath + " is open.")

    def setPitch(self, value):

        self.pitch = value;
        print("The pitch value is: " + str(self.pitch))

#Winamp concrete class
class Winamp(Player):


    def __init__(self):

        print("The Winamp constructor was invoked.")

    def __del__(self):

        print("The Winamp destructor was invoked.")

    def open(self, filePath):

        #@ToDo: por ahora simulamos la apertura correcta
        self.isOpen = True;
        print("The audiofile: " + filePath + " is open.")



    #NO SE PUEDE CREAR UN OBJETO O INSTANCIA DE UNA CLASE ABSTRACTA:
    #Player player;//NO!

vlcPlayer = VLC()
vlcPlayer.open("./resources/orchestral.ogg");
vlcPlayer.play();
vlcPlayer.setVolume(13);


winampPlayer = Winamp()

winampPlayer.open("./resources/orchestral.ogg");
winampPlayer.play();
winampPlayer.setVolume(13);

